File: src/crewlet/db/secret_values.py
"""Typed queries against the ``secret_values`` table — the secret store.

One row per ``${VAR}`` name, each value sealed with the Tier A keyring.
:class:`SecretValueStore` is the read/write surface;
:class:`DatabaseSecretSource` is the boot-loaded snapshot that
``crewlet.config._resolve_env_value`` consults ahead of ``os.environ``.

Unlike :class:`~crewlet.db.company_config.CompanyConfigStore`, this store
takes the cipher itself rather than accepting pre-sealed payloads.  That
is deliberate: each row binds its own ``name`` as AEAD associated data, so
leaving the sealing to callers would spread one easily-mismatched
convention across every write site.  One place owns the AAD, and a
ciphertext lifted into another row simply fails to decrypt.

See ``docs/concepts/secret-store.md``.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Any

from crewlet._logging import get_logger
from crewlet.db.client import Database
from crewlet.env_refs import is_env_var_name
from crewlet.secrets.cipher import SecretCipher, parse_key_id

logger = get_logger("db.secret_values")

#: Associated-data domain for a stored secret. The row's ``name`` is bound
#: in, so a ciphertext moved between rows fails to decrypt instead of
#: silently answering as a different secret.
_AAD_PREFIX = "secret_values/"


def _aad(name: str) -> str:
    return f"{_AAD_PREFIX}{name}"


class SecretStoreError(Exception):
    """The secret store cannot serve a request."""


@dataclass
class SecretValueRecord:
    """Metadata for one stored secret — deliberately **without** the value.

    Listing secrets is an inventory operation; a record type that cannot
    carry a value is what keeps ``crewlet secrets list`` and any future
    read path structurally incapable of leaking one.
    """

    name: str
    key_id: str
    updated_at: datetime
    updated_by: str
    source: str


def _row_to_record(row: dict[str, Any]) -> SecretValueRecord:
    return SecretValueRecord(
        name=row["name"],
        key_id=row["key_id"],
        updated_at=row["updated_at"],
        updated_by=row["updated_by"],
        source=row["source"],
    )


class SecretValueStore:
    """Persist and query encrypted secret values keyed by var name.

    A keyring is **required**: every method that touches a value raises
    :class:`SecretStoreError` when ``cipher is None``.  Failing loudly
    beats the alternative — a missing keyring silently yielding ``""``
    produces an empty Bearer token that looks live to every layer until
    the provider rejects it, hours later and far from the cause.
    """

    def __init__(self, db: Database, cipher: SecretCipher | None) -> None:
        self._db = db
        self._cipher = cipher

    @property
    def enabled(self) -> bool:
        """True when a keyring is configured, so values can be sealed."""
        return self._cipher is not None

    def _require_cipher(self) -> SecretCipher:
        if self._cipher is None:
            raise SecretStoreError(
                "the secret store needs an encryption keyring — set Tier A "
                "`secrets.keys` in config.yaml (generate one with "
                "`crewlet secrets keygen`). Secrets are never stored "
                "unencrypted."
            )
        return self._cipher

    async def put(
        self,
        name: str,
        value: str,
        *,
        updated_by: str,
        source: str,
    ) -> None:
        """Seal ``value`` under ``name``, inserting or replacing the row."""
        if not is_env_var_name(name):
            raise SecretStoreError(
                f"{name!r} is not a valid environment-variable name, so no "
                f"${{...}} reference could ever read it "
                f"(expected [A-Za-z_][A-Za-z0-9_]*)"
            )
        cipher = self._require_cipher()
        sealed = cipher.encrypt(value, aad=_aad(name))
        await self._db.execute(
            """
            INSERT INTO secret_values (name, value, key_id, updated_by, source)
            VALUES ($1, $2, $3, $4, $5)
            ON CONFLICT (name) DO UPDATE
               SET value = EXCLUDED.value,
                   key_id = EXCLUDED.key_id,
                   updated_at = now(),
                   updated_by = EXCLUDED.updated_by,
                   source = EXCLUDED.source
            """,
            name,
            sealed,
            parse_key_id(sealed) or "",
            updated_by,
            source,
        )
        # Name and provenance only — the value never reaches a log line.
        logger.info("secret_value_stored", name=name, source=source)

    async def get(self, name: str) -> str | None:
        """The decrypted value for ``name``, or ``None`` when not stored."""
        row = await self._db.fetchrow(
            "SELECT value FROM secret_values WHERE name = $1", name
        )
        if row is None:
            return None
        cipher = self._require_cipher()
        return cipher.decrypt(row["value"], aad=_aad(name))

    async def delete(self, name: str) -> bool:
        """Remove ``name``. Returns whether a row was actually removed."""
        rows = await self._db.execute(
            "DELETE FROM secret_values WHERE name = $1 RETURNING name", name
        )
        removed = bool(rows)
        if removed:
            logger.info("secret_value_deleted", name=name)
        return removed

    async def list_records(self) -> list[SecretValueRecord]:
        """Metadata for every stored secret, newest write first.

        Needs no keyring — an operator locked out of the key can still see
        what is stored and under which ``key_id``, which is exactly the
        inventory a key-loss postmortem needs.
        """
        rows = await self._db.execute(
            """
            SELECT name, key_id, updated_at, updated_by, source
            FROM secret_values
            ORDER BY updated_at DESC
            """
        )
        return [_row_to_record(row) for row in rows]

    async def load_all(self) -> dict[str, str]:
        """Every stored secret, decrypted — the process snapshot.

        Returns ``{}`` for an empty table even with no keyring, so a
        deployment that has never used the store stays inert.  Rows with
        no keyring raise instead: answering ``""`` for a name someone
        deliberately stored is the failure mode this store exists to
        prevent.
        """
        rows = await self._db.execute("SELECT name, value FROM secret_values")
        if not rows:
            return {}
        cipher = self._require_cipher()
        return {
            row["name"]: cipher.decrypt(row["value"], aad=_aad(row["name"]))
            for row in rows
        }

    async def rekey(self) -> list[str]:
        """Re-seal every row not already under the active key.

        The per-row counterpart of ``crewlet config rekey``: without it a
        retired key could never be dropped from the keyring, because rows
        sealed under it would become unreadable.  Returns the names
        re-sealed.
        """
        cipher = self._require_cipher()
        active = getattr(cipher, "active_key_id", "")
        rows = await self._db.execute("SELECT name, value, key_id FROM secret_values")
        rotated: list[str] = []
        for row in rows:
            if row["key_id"] == active:
                continue
            name = row["name"]
            plaintext = cipher.decrypt(row["value"], aad=_aad(name))
            sealed = cipher.encrypt(plaintext, aad=_aad(name))
            await self._db.execute(
                """
                UPDATE secret_values
                   SET value = $2, key_id = $3, updated_at = now()
                 WHERE name = $1
                """,
                name,
                sealed,
                parse_key_id(sealed) or "",
            )
            rotated.append(name)
        if rotated:
            logger.info("secret_values_rekeyed", names=rotated, key_id=active)
        return rotated


class DatabaseSecretSource:
    """A refreshable in-memory snapshot of :class:`SecretValueStore`.

    Implements :class:`crewlet.secrets.resolver.SecretSource`.  Resolution
    is synchronous at ~25 scattered call sites, so the values are read
    once into a dict and swapped wholesale on refresh — never queried
    per lookup.

    The swap is atomic (one rebind of ``_values``), so a refresh racing a
    resolution yields either the old map or the new one, never a
    half-updated one.
    """

    def __init__(self, store: SecretValueStore) -> None:
        self._store = store
        self._values: dict[str, str] = {}

    def get(self, name: str) -> str | None:
        return self._values.get(name)

    def names(self) -> list[str]:
        """Stored names — used to report what the store shadows in env."""
        return sorted(self._values)

    async def refresh(self) -> None:
        """Re-read the table and swap the snapshot in."""
        loaded = await self._store.load_all()
        previous = set(self._values)
        self._values = loaded
        if set(loaded) != previous:
            logger.info("secret_snapshot_refreshed", count=len(loaded))


async def load_secret_source(
    db: Database, cipher: SecretCipher | None
) -> DatabaseSecretSource | None:
    """Build and prime a snapshot, or ``None`` when the store is unused.

    ``None`` for an empty table keeps every existing deployment on the
    exact ``os.environ`` path it has today — the feature engages only once
    a secret is actually stored.  A populated table with no keyring raises
    :class:`SecretStoreError` rather than booting with secrets it cannot
    read.
    """
    source = DatabaseSecretSource(SecretValueStore(db, cipher))
    await source.refresh()
    if not source.names():
        return None
    logger.info("secret_store_loaded", count=len(source.names()))
    return source


__all__ = [
    "DatabaseSecretSource",
    "SecretStoreError",
    "SecretValueRecord",
    "SecretValueStore",
    "load_secret_source",
]

File: internal/store/secretvalues.go
package store

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"sort"
	"time"

	"github.com/crewlet/crewlet/internal/secrets"
)

// SecretValues is this node's own encrypted secret rows, keyed by the ${VAR}
// name a config value references.
//
// # It is the BOOTSTRAP and MIGRATION store, not the fleet's
//
// The company's secrets live on the coordination KV, where every node reads
// them — see [github.com/crewlet/crewlet/internal/fleetsecrets] and
// decisions/203. This table is what a node can write when the KV is out of
// reach, which on the default embedded topology is any moment the engine is
// not running: the broker is in the engine's own process, so `crewlet secrets
// set` on a stopped node has nowhere else to put a value.
//
// The engine MIGRATES these rows onto the fleet at its next start and removes
// them here, which is what stops a stale local row shadowing a rotation
// forever. Until that start they are this node's alone, and no peer can see
// them.
//
// # It is the companion to company_config, not a duplicate of it
//
// Rotation is an UPDATE of one row here. Writing the literal into the config
// instead would archive the OLD secret forever, because every revision is an
// immutable copy and revisions are never scrubbed. And one name can have
// many pointers: a bot token referenced from an integration block and from a
// per-role MCP env is ONE credential with two readers, which keying by var
// name keeps as one row.
//
// # No plaintext mode, unlike company_config
//
// A keyring is REQUIRED to read or write a row. There is no legacy corpus to
// stay compatible with, and a dedicated secret store that can hold
// unencrypted secrets is a footgun with no upside.
//
// # Reads fail CLOSED
//
// Everything else in this package that cannot reach its data returns the
// safe answer and carries on. This one raises. An unreadable secret
// resolving to "" does not fail here — it becomes an empty Bearer token
// hours later and somewhere else entirely, on a request whose 401 names the
// vendor rather than this store.
type SecretValues struct {
	db     *DB
	cipher secrets.Cipher
}

// SecretValues returns the secret store sealed with this cipher.
//
// A NIL CIPHER IS A STORE THAT REFUSES EVERYTHING rather than one that
// stores plaintext — see the type comment. It is not an error to construct,
// because a node with no keyring is a supported deployment; it is an error
// to use.
func (d *DB) SecretValues(cipher secrets.Cipher) *SecretValues {
	return &SecretValues{db: d, cipher: cipher}
}

const secretUpsertSQL = `
INSERT INTO secret_values (name, value, key_id, updated_at, updated_by, source)
VALUES (?, ?, ?, ?, ?, ?)
ON CONFLICT (name) DO UPDATE SET
    value = excluded.value, key_id = excluded.key_id,
    updated_at = excluded.updated_at, updated_by = excluded.updated_by,
    source = excluded.source`

// Set seals a secret and stores it under name, replacing any prior value.
//
// AN UPSERT rather than a read-then-write, because rotation is the common
// path and two operators rotating at once must not produce a row that holds
// neither of their values.
func (s *SecretValues) Set(ctx context.Context, name, value, by, source string, now time.Time) error {
	if s.cipher == nil {
		return secrets.ErrNoKeyring
	}
	if name == "" {
		return errors.New("store: a secret needs a name")
	}
	// THE NAME IS BOUND IN as associated data, so a ciphertext moved to
	// another row fails to decrypt instead of silently impersonating a
	// different secret — which is what an attacker with UPDATE but not the
	// key would otherwise be able to do.
	sealed, err := s.cipher.Encrypt(value, secrets.AADForVar(name))
	if err != nil {
		return fmt.Errorf("store: seal %s: %w", name, err)
	}
	keyID, ok := secrets.EnvelopeKeyID(sealed)
	if !ok {
		return fmt.Errorf("store: seal %s: the cipher produced no key id", name)
	}
	_, err = s.db.sql.ExecContext(ctx, secretUpsertSQL, name, sealed, keyID,
		EncodeTime(now), by, source)
	if err != nil {
		return fmt.Errorf("store: write secret %s: %w", name, err)
	}
	return nil
}

// Get unseals one secret.
//
// The error says WHICH failure it was — see [secrets.ErrNotFound] — and never
// carries the value or the ciphertext: a decrypt error that echoed its input
// would put an envelope into a log that a keyring might later open.
func (s *SecretValues) Get(ctx context.Context, name string) (string, error) {
	if s.cipher == nil {
		return "", secrets.ErrNoKeyring
	}
	var sealed string
	err := s.db.sql.QueryRowContext(ctx,
		`SELECT value FROM secret_values WHERE name = ?`, name).Scan(&sealed)
	if errors.Is(err, sql.ErrNoRows) {
		return "", fmt.Errorf("%w: %s", secrets.ErrNotFound, name)
	}
	if err != nil {
		return "", fmt.Errorf("store: read secret %s: %w", name, err)
	}
	plain, err := s.cipher.Decrypt(sealed, secrets.AADForVar(name))
	if err != nil {
		return "", fmt.Errorf("store: open secret %s: %w", name, err)
	}
	return plain, nil
}

// List returns every stored secret's metadata, name-ordered.
//
// NAME-ORDERED rather than by recency, because this is what an operator
// reads to answer "is X set", and a list that moves every time something is
// rotated is one they have to search rather than scan.
//
// NO VALUES: [secrets.Record.Value] is left blank, and the column is not even
// selected. A listing is printed, and a listing that carried envelopes would
// put every credential a company has into one scrollback buffer.
func (s *SecretValues) List(ctx context.Context) ([]secrets.Record, error) {
	rows, err := s.db.sql.QueryContext(ctx,
		`SELECT name, key_id, updated_at, updated_by, source
		 FROM secret_values ORDER BY name`)
	if err != nil {
		return nil, fmt.Errorf("store: list secrets: %w", err)
	}
	defer rows.Close()
	var out []secrets.Record
	for rows.Next() {
		var r secrets.Record
		var micros int64
		if err := rows.Scan(&r.Name, &r.KeyID, &micros, &r.UpdatedBy, &r.Source); err != nil {
			return nil, fmt.Errorf("store: scan secret: %w", err)
		}
		r.UpdatedAt = DecodeTime(micros)
		out = append(out, r)
	}
	return out, rows.Err()
}

// Unset removes a secret, reporting whether a row went.
func (s *SecretValues) Unset(ctx context.Context, name string) (bool, error) {
	res, err := s.db.sql.ExecContext(ctx,
		`DELETE FROM secret_values WHERE name = ?`, name)
	if err != nil {
		return false, fmt.Errorf("store: unset secret %s: %w", name, err)
	}
	n, err := res.RowsAffected()
	if err != nil {
		return false, fmt.Errorf("store: unset secret %s: %w", name, err)
	}
	return n > 0, nil
}

// All unseals every secret, for the boot snapshot the resolver reads.
//
// A SNAPSHOT rather than a per-lookup query, because ${VAR} resolution
// happens deep inside config expansion — per role, per provider, per MCP
// server — and a database round trip there would put the store on the path
// of every config read, including the ones a validate command makes with no
// database at all.
//
// It FAILS CLOSED on the first row it cannot open: a partial snapshot is
// the worst outcome, because the names that are missing resolve to whatever
// the environment happens to hold, which is exactly the stale-.env shadowing
// the store exists to prevent.
func (s *SecretValues) All(ctx context.Context) (map[string]string, error) {
	if s.cipher == nil {
		return nil, secrets.ErrNoKeyring
	}
	rows, err := s.db.sql.QueryContext(ctx,
		`SELECT name, value FROM secret_values ORDER BY name`)
	if err != nil {
		return nil, fmt.Errorf("store: read secrets: %w", err)
	}
	defer rows.Close()
	out := map[string]string{}
	for rows.Next() {
		var name, sealed string
		if err := rows.Scan(&name, &sealed); err != nil {
			return nil, fmt.Errorf("store: scan secret: %w", err)
		}
		plain, err := s.cipher.Decrypt(sealed, secrets.AADForVar(name))
		if err != nil {
			return nil, fmt.Errorf("store: open secret %s: %w", name, err)
		}
		out[name] = plain
	}
	return out, rows.Err()
}

// Rekey re-seals every row that is not already under the keyring's active
// key, reporting the names it moved.
//
// NAMES RATHER THAN A COUNT, because that is what an operator needs to
// confirm afterwards, and because a rekey that moved 12 of 13 rows is a
// question ("which one?") that a count cannot answer.
//
// One transaction: a rekey interrupted halfway leaves rows under a key the
// operator is about to retire, and the whole point of the pass is being able
// to retire it.
func (s *SecretValues) Rekey(ctx context.Context, activeKeyID, by string, now time.Time) ([]string, error) {
	if s.cipher == nil {
		return nil, secrets.ErrNoKeyring
	}
	var moved []string
	err := s.db.Tx(ctx, func(tx *sql.Tx) error {
		rows, err := tx.QueryContext(ctx,
			`SELECT name, value FROM secret_values WHERE key_id != ? ORDER BY name`,
			activeKeyID)
		if err != nil {
			return fmt.Errorf("store: find stale secrets: %w", err)
		}
		// Collected before writing: the driver holds one connection per
		// transaction, so writing while the rows are open would deadlock
		// against the read this is iterating.
		stale := map[string]string{}
		for rows.Next() {
			var name, sealed string
			if err := rows.Scan(&name, &sealed); err != nil {
				rows.Close()
				return fmt.Errorf("store: scan secret: %w", err)
			}
			stale[name] = sealed
		}
		rows.Close()
		if err := rows.Err(); err != nil {
			return fmt.Errorf("store: find stale secrets: %w", err)
		}

		for _, name := range sortedNames(stale) {
			plain, err := s.cipher.Decrypt(stale[name], secrets.AADForVar(name))
			if err != nil {
				// ABORTS THE WHOLE PASS. A row this keyring cannot open
				// is a key that was dropped from the config, and moving
				// the others while leaving it would report a successful
				// rekey over a secret that is now unreadable for ever.
				return fmt.Errorf("store: open secret %s for rekey: %w", name, err)
			}
			sealed, err := s.cipher.Encrypt(plain, secrets.AADForVar(name))
			if err != nil {
				return fmt.Errorf("store: re-seal %s: %w", name, err)
			}
			if _, err := tx.ExecContext(ctx, secretUpsertSQL, name, sealed,
				activeKeyID, EncodeTime(now), by, "rekey"); err != nil {
				return fmt.Errorf("store: write re-sealed %s: %w", name, err)
			}
			moved = append(moved, name)
		}
		return nil
	})
	if err != nil {
		return nil, err
	}
	return moved, nil
}

func sortedNames(m map[string]string) []string {
	out := make([]string, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}
